torch/distributed/_spmd/api.py

```diff
@@ -51,3 +73,119 @@ def forward(self, *args: Tuple[object], **kwargs: Dict[str, object]) -> object:
 
         assert self._compiled_m is not None
         return self._compiled_m(*args, **kwargs)
+
+
+class Override(ABC):
+    @abstractmethod
+    def replacement(self, orig_submodule: torch.nn.Module) -> torch.nn.Module:
+        pass
+
+    @abstractmethod
+    def transform(
+        self, gm: fx.GraphModule, context: Dict[fx.Node, DTensor]
+    ) -> fx.Graph:
+        pass
+
+
+OVERRIDES: Dict[Type[Any], Override] = {}
+
+
+def _register_override(typ: Any, override: Override) -> None:
+    OVERRIDES[typ] = override
+
+
+def _dtensor_expand(gm, args, kwargs, params_and_buffers):
+    flat_args, _ = pytree.tree_flatten(list(args) + list(kwargs.values()))
+
+    mesh = DeviceMesh("cuda", torch.arange(dist.get_world_size()).cuda())
+    shard_schema: Schema = Schema(mesh=mesh, placements=[Shard(0)])
+    replicate_schema: Schema = Schema(mesh=mesh, placements=[Replicate()])
+
+    inps = []
+    schemas = []
+    for a in flat_args:
+        if isinstance(a, torch.Tensor):
+            inps.append(a)
+            schemas.append(shard_schema)
+        elif isinstance(a, nn.Module):
+            # nn.Module placeholder is captured by make_fx but never used in the graph
+            inps.append(torch.empty(0))
+            schemas.append(shard_schema)
+
+    flat_params, _ = pytree.tree_flatten(params_and_buffers)
+
+    for p in flat_params:
+        inps.append(p)
+        schemas.append(replicate_schema)
+
+    # fake_mode: FakeTensorMode = FakeTensorMode()
+    # fake_inps = [fake_mode.from_tensor(inp) for inp in inps]
+
+    return _convert_to_distributed(gm, inps, schemas, _allow_partial=False)[0]
+
+
+def _compile():
+    def inner(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            # 1. Extract nn.Module from args and kwargs
+            # FIXME(@mrshenli): detect nn.Module in a nested way
+            # FIXME(@mrshenli): support multiple nn.Module instances
+            # FIXME(@mrshenli): do the same for optimizers
```

Comment:
Can you add optimizer into the example? IMO the purpose of capturing the train step as opposed to just the model fwd/bwd is to capture optimizers. It's not yet clear how things comes together here.